src/System.Text.Json/src/System/Text/Json/Serialization/Converters/DefaultImmutableConverter.cs

```diff
@@ -0,0 +1,142 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Collections;
+using System.Collections.Generic;
+using System.Collections.Concurrent;
+using System.Diagnostics;
+using System.Text.Json.Serialization.Policies;
+
+namespace System.Text.Json.Serialization.Converters
+{
+    // This converter returns enumerables in the System.Collections.Immutable namespace.
+    internal sealed class DefaultImmutableConverter : JsonEnumerableConverter
+    {
+        public const string ImmutableNamespace = "System.Collections.Immutable";
+
+        private const string ImmutableListTypeName = "System.Collections.Immutable.ImmutableList";
+        private const string ImmutableListGenericTypeName = "System.Collections.Immutable.ImmutableList`1";
+        private const string ImmutableListGenericInterfaceTypeName = "System.Collections.Immutable.IImmutableList`1";
+
+        private const string ImmutableStackTypeName = "System.Collections.Immutable.ImmutableStack";
+        private const string ImmutableStackGenericTypeName = "System.Collections.Immutable.ImmutableStack`1";
+        private const string ImmutableStackGenericInterfaceTypeName = "System.Collections.Immutable.IImmutableStack`1";
+
+        private const string ImmutableQueueTypeName = "System.Collections.Immutable.ImmutableQueue";
+        private const string ImmutableQueueGenericTypeName = "System.Collections.Immutable.ImmutableQueue`1";
+        private const string ImmutableQueueGenericInterfaceTypeName = "System.Collections.Immutable.IImmutableQueue`1";
+
+        private const string ImmutableSortedSetTypeName = "System.Collections.Immutable.ImmutableSortedSet";
+        private const string ImmutableSortedSetGenericTypeName = "System.Collections.Immutable.ImmutableSortedSet`1";
+
+        private const string ImmutableHashSetTypeName = "System.Collections.Immutable.ImmutableHashSet";
+        private const string ImmutableHashSetGenericTypeName = "System.Collections.Immutable.ImmutableHashSet`1";
+        private const string ImmutableSetGenericInterfaceTypeName = "System.Collections.Immutable.IImmutableSet`1";
+
+        internal delegate object ImmutableCreateRangeDelegate<T>(IEnumerable<T> items);
+        internal delegate IEnumerable CreateImmutableCollectionDelegate(string createRangeDelegateKey, IList sourceList);
+
+        private static readonly ConcurrentDictionary<string, CreateImmutableCollectionDelegate> _CreateImmutableCollectionDelegates = new ConcurrentDictionary<string, CreateImmutableCollectionDelegate>();
+
+        private static ConcurrentDictionary<string, object> _createRangeDelegates = new ConcurrentDictionary<string, object>();
+
+        private static readonly ConcurrentDictionary<Type, string> _createRangeDelegateKeys = new ConcurrentDictionary<Type, string>();
```

Comment:
Can you simplify this down to a single concurrent dictionary using `ConcurrentDictionary<Type, (CreateImmutableCollectionDelegate, object)>`?

This will require less memory and fewer lookups.

Reply:
I got this down to two dictionaries. I need one to map "constructing types" to their `CreateRange` delegates, and one to map element types to `CreateImmutable` delegates.

If we do a single `ConcurrentDictionary<Type, (CreateImmutableCollectionDelegate, object)>` we'd have to repeat the `CreateImmutableDelegate` construction for element types _every_ time we see some `(I)ImmutableX<T>` (e.g once when we `ImmutableStack<string>`, again when we see `IImmutableList<string>` etc. vs only once when we see `ImmutableStack<string>`). I can explain more offline if this is confusing.

Reply:
That makes sense, thanks 👍